Upgrade Tabulator to 5.x

philippjfr opened the pull request:
Tabulator 5.0 is a major upgrade to the tabulator.js library which applied some major refactoring and improves performance. Since we are still very actively improving the Tabulator widget in Panel and don't want to invest in developing features on top of an old version and also want to leverage the latest in performance and features it is best to track tabulator.js as closely as possible.

Closes https://github.com/holoviz/panel/issues/2832

- [x] Fix pagination
- [x] Fix frozen rows

maximlt commented on panel/models/tabulator.py:
```diff
@@ -15,10 +15,10 @@
-JS_SRC = "https://unpkg.com/tabulator-tables@4.9.3/dist/js/tabulator.js"
-MOMENT_SRC = "https://unpkg.com/moment@2.27.0/moment.js"
+JS_SRC = "https://unpkg.com/tabulator-tables@5.0.7/dist/js/tabulator.js"
+MOMENT_SRC = "https://cdn.jsdelivr.net/npm/luxon/build/global/luxon.min.js"
```
Shouldn't this be pinned somehow?

philippjfr replied:
Luxon is the replacement for moment.js which is THE datetime library for JS. So there will be no API changes here as that would break entire ecosystems.

maximlt commented:
I get an error when I hit Enter on the example below:
![issue1](https://user-images.githubusercontent.com/35924738/140972433-b68f5a0d-91a8-4b37-8cbf-50335772e7d4.gif)

maximlt commented:
The cell content of every second date row is made invisible when I click on the cell in this example:
![issue2](https://user-images.githubusercontent.com/35924738/140972775-01e2a0be-48f4-4bb9-8de9-fa265aed94cb.gif)

maximlt commented:
Setting `selectable` to an integer (maximum number of rows selectable) with for instance `pn.widgets.Tabulator(sel_df, selectable=4)` raises the following error when a row is clicked (it can even be selected):
```
18:26:50.962 Uncaught TypeError: this.model.select_mode.startsWith is not a function
    rowSelectionChanged Tabulator.ipynb line 2 > injectedScript:1511
    init_callbacks Tabulator.ipynb line 2 > injectedScript:1511
    _dispatch ExternalEventBus.js:82
    _dispatch ExternalEventBus.js:81
    dispatchExternal CoreFeature.js:90
    _rowSelectionChanged SelectRow.js:402
    deselectRows SelectRow.js:310
    rowClicked Tabulator.ipynb line 2 > injectedScript:1511
    init_callbacks Tabulator.ipynb line 2 > injectedScript:1511
    _dispatch ExternalEventBus.js:82
    _dispatch ExternalEventBus.js:81
    dispatchExternal CoreFeature.js:90
    dispatchEvent Interaction.js:303
    handle Interaction.js:233
    _dispatch InternalEventBus.js:119
    _dispatch InternalEventBus.js:118
    dispatch CoreFeature.js:78
    triggerEvents InteractionMonitor.js:206
    track InteractionMonitor.js:112
    updateEventListeners InteractionMonitor.js:96
    subscriptionChanged InteractionMonitor.js:85
    _notifiySubscriptionChange InternalEventBus.js:108
    _notifiySubscriptionChange InternalEventBus.js:107
    subscribe InternalEventBus.js:36
    subscribe CoreFeature.js:62
    subscriptionChanged Interaction.js:159
    _notifiySubscriptionChange ExternalEventBus.js:70
    _notifiySubscriptionChange ExternalEventBus.js:69
    subscriptionChange ExternalEventBus.js:21
    subscriptionChangeExternal CoreFeature.js:98
    initializeExternalEvents Interaction.js:148
    initialize Interaction.js:117
    _initializeTable Tabulator.js:277
    _create Tabulator.js:139
    Tabulator Tabulator.js:57
    setTimeout handler*Tabulator Tabulator.js:56
    _createSuperInternal tabulator.js:116
    TabulatorFull tabulator.js:24950
    render Tabulator.ipynb line 2 > injectedScript:1511
    build Tabulator.ipynb line 2 > injectedScript:674
    renderTo Tabulator.ipynb line 2 > injectedScript:674
    f Tabulator.ipynb line 2 > injectedScript:762
    add_document_standalone Tabulator.ipynb line 2 > injectedScript:762
    embed_items_notebook Tabulator.ipynb line 2 > injectedScript:769
    embed_document Tabulator.ipynb line 2 > injectedScript:5
    <anonymous> Tabulator.ipynb line 2 > injectedScript:8
    <anonymous> Tabulator.ipynb line 2 > injectedScript:24
    attachWidget jlab_core.1ad82bcd8c48e07a8392.js:2
    insertWidget jlab_core.1ad82bcd8c48e07a8392.js:2
    _insertOutput jlab_core.1ad82bcd8c48e07a8392.js:2
    onModelChanged jlab_core.1ad82bcd8c48e07a8392.js:2
    m jlab_core.1ad82bcd8c48e07a8392.js:2
    l jlab_core.1ad82bcd8c48e07a8392.js:2
    emit jlab_core.1ad82bcd8c48e07a8392.js:2
    _onListChanged jlab_core.1ad82bcd8c48e07a8392.js:2
    m jlab_core.1ad82bcd8c48e07a8392.js:2
    l jlab_core.1ad82bcd8c48e07a8392.js:2
    emit jlab_core.1ad82bcd8c48e07a8392.js:2
    push jlab_core.1ad82bcd8c48e07a8392.js:2
    _add jlab_core.1ad82bcd8c48e07a8392.js:2
    add jlab_core.1ad82bcd8c48e07a8392.js:2
    _onIOPub jlab_core.1ad82bcd8c48e07a8392.js:2
    _handleIOPub jlab_core.1ad82bcd8c48e07a8392.js:2
Tabulator.ipynb line 2 > injectedScript:1511:13698
```

maximlt commented:
When moving the horizontal scroller from left to right the frozen column moves a little:
![issue4](https://user-images.githubusercontent.com/35924738/140974634-17e557e3-7e0c-43b1-8f84-6447cbb73436.gif)

maximlt commented:
I get an error in the console when I click on a row (not the `+` but the row itself as if I was about to select it) on this example:
![image](https://user-images.githubusercontent.com/35924738/140975321-e1d983c1-94c7-43d4-a910-07581738b41f.png)

```
18:33:49.179 Uncaught Error: Selection(3241).indices given invalid value: [null]
    validate Tabulator.ipynb line 2 > injectedScript:361
    _update Tabulator.ipynb line 2 > injectedScript:361
    set_value Tabulator.ipynb line 2 > injectedScript:361
    _setv Tabulator.ipynb line 2 > injectedScript:357
    setv Tabulator.ipynb line 2 > injectedScript:357
    set Tabulator.ipynb line 2 > injectedScript:357
    rowClicked Tabulator.ipynb line 2 > injectedScript:1511
    init_callbacks Tabulator.ipynb line 2 > injectedScript:1511
    _dispatch ExternalEventBus.js:82
    _dispatch ExternalEventBus.js:81
    dispatchExternal CoreFeature.js:90
    dispatchEvent Interaction.js:303
    handle Interaction.js:233
    _dispatch InternalEventBus.js:119
    _dispatch InternalEventBus.js:118
    dispatch CoreFeature.js:78
    triggerEvents InteractionMonitor.js:206
    track InteractionMonitor.js:112
    updateEventListeners InteractionMonitor.js:96
    subscriptionChanged InteractionMonitor.js:85
    _notifiySubscriptionChange InternalEventBus.js:108
    _notifiySubscriptionChange InternalEventBus.js:107
    subscribe InternalEventBus.js:36
    subscribe CoreFeature.js:62
    subscriptionChanged Interaction.js:159
    _notifiySubscriptionChange ExternalEventBus.js:70
    _notifiySubscriptionChange ExternalEventBus.js:69
    subscriptionChange ExternalEventBus.js:21
    subscriptionChangeExternal CoreFeature.js:98
    initializeExternalEvents Interaction.js:148
    initialize Interaction.js:117
    _initializeTable Tabulator.js:277
    _create Tabulator.js:139
    Tabulator Tabulator.js:57
    setTimeout handler*Tabulator Tabulator.js:56
    _createSuperInternal tabulator.js:116
    TabulatorFull tabulator.js:24950
    render Tabulator.ipynb line 2 > injectedScript:1511
    build Tabulator.ipynb line 2 > injectedScript:674
    renderTo Tabulator.ipynb line 2 > injectedScript:674
    f Tabulator.ipynb line 2 > injectedScript:762
    add_document_standalone Tabulator.ipynb line 2 > injectedScript:762
    embed_items_notebook Tabulator.ipynb line 2 > injectedScript:769
    embed_document Tabulator.ipynb line 2 > injectedScript:5
    <anonymous> Tabulator.ipynb line 2 > injectedScript:8
    <anonymous> Tabulator.ipynb line 2 > injectedScript:24
    attachWidget jlab_core.1ad82bcd8c48e07a8392.js:2
    insertWidget jlab_core.1ad82bcd8c48e07a8392.js:2
    _insertOutput jlab_core.1ad82bcd8c48e07a8392.js:2
    onModelChanged jlab_core.1ad82bcd8c48e07a8392.js:2
    m jlab_core.1ad82bcd8c48e07a8392.js:2
    l jlab_core.1ad82bcd8c48e07a8392.js:2
    emit jlab_core.1ad82bcd8c48e07a8392.js:2
    _onListChanged jlab_core.1ad82bcd8c48e07a8392.js:2
    m jlab_core.1ad82bcd8c48e07a8392.js:2
    l jlab_core.1ad82bcd8c48e07a8392.js:2
    emit jlab_core.1ad82bcd8c48e07a8392.js:2
    push jlab_core.1ad82bcd8c48e07a8392.js:2
    _add jlab_core.1ad82bcd8c48e07a8392.js:2
    add jlab_core.1ad82bcd8c48e07a8392.js:2
    _onIOPub jlab_core.1ad82bcd8c48e07a8392.js:2
Tabulator.ipynb line 2 > injectedScript:361:2552

```

maximlt commented:
When I select some rows on a page of a paginated table and then go to another page I get an error in the console:
![issue5](https://user-images.githubusercontent.com/35924738/140975655-e3807210-9883-455b-8167-29f17de35862.gif)

```
18:37:10.682 Uncaught (in promise) Scroll Error - No matching row found Tabulator.js:663:18
```

I cannot select rows that are on different pages (here I'm holding Ctrl when clicking on the rows of the 2nd page, to add more rows to the selection):
![issue6](https://user-images.githubusercontent.com/35924738/140976112-ec46926b-40ed-4d6d-ad59-c9c2c0c41da9.gif)

Also note that `.selected_dataframe` seems to be equal to the whole dataframe at the initialization.

maximlt commented:
There are still issues with the client autocomplete filter. Here I cannot reset the table after deleting `A`:
![issue7](https://user-images.githubusercontent.com/35924738/140976998-de1a2cde-0f65-449f-b0fa-ec0fe3159d6f.gif)

philippjfr commented:
Thanks, everything I'm not replying to below I'll look into and fix in this PR.

> I get an error when I hit Enter on the example below: ![issue1](https://user-images.githubusercontent.com/35924738/140972433-b68f5a0d-91a8-4b37-8cbf-50335772e7d4.gif)

Fixed in follow up PR for datetime editing.

> The cell content of every second date row is made invisible when I click on the cell in this example: ![issue2](https://user-images.githubusercontent.com/35924738/140972775-01e2a0be-48f4-4bb9-8de9-fa265aed94cb.gif)

Also addressed by the follow up PR, but this didn't work before in any case.

> When moving the horizontal scroller from left to right the frozen column moves a little: ![issue4](https://user-images.githubusercontent.com/35924738/140974634-17e557e3-7e0c-43b1-8f84-6447cbb73436.gif)

Not really our fault. Hopefully will get addressed in Tabulator.

>There are still issues with the client autocomplete filter. Here I cannot reset the table after deleting A:

Unrelated to this upgrade, so ignoring this for now.

philippjfr commented:
> I get an error in the console when I click on a row (not the `+` but the row itself as if I was about to select it) on this example: ![image](https://user-images.githubusercontent.com/35924738/140975321-e1d983c1-94c7-43d4-a910-07581738b41f.png)

Perfectly okay saying that selections won't work on hierarchical tables but I'll add some validation for this.

philippjfr commented:
Okay fixed three of the issues mentioned:

- No more errors on scroll to row if selected rows aren't on current page
- Selections disabled on hierarchical tables
- No more errors when setting selectable to integer value

I don't think any of these were caused by this PR but good to have them fixed anyway. The datetime editor PR will address some of the other issues.
